### PR TITLE
TOOLS: various script fixes, VISION: add RectFinder

### PR DESCRIPTION
**Pull request by kev-the-dev**
* Improve label me bag interface. Each segment can now contain multiple image topics (ex: stereo) and will produce its own bag (rather than one bag for every segment produced from the same input bag)
* fix bag has a few new options, including specifying timestamps to crop a bag and a list of topics to either include or ignore in the output bag

I tested the labelme script with the bags in the path marker and it worked well. I gave it a YAML, ran the script, and the images showed up on my LabelMe server. After I finished labeling, I ran it with the ```-i``` flag and it produced a bag for each segment containing the original data and the labels.

### CODE REVIEW

File: utils/mil_tools/mil_ros_tools/label_me_bag.py [hunk 1]
```diff
@@ -72,21 +75,22 @@ def verify_attr(obj, attr, objname, throw=True):
...
-          for segment in bag['segments']:
-              verify_attr(segment, 'start', 'segment', throw=False)
-              verify_attr(segment, 'stop', 'segment', throw=False)
-              verify_attr(segment, 'freq', 'segment', throw=False)
-              verify_attr(segment, 'topic', 'segment')
+          for j, segment in enumerate(bag['segments']):
+              verify_attr(segment, 'topics', 'segment')
+              # Ensure topics is still a list
+              if type(segment['topics']) == str:
+                  self.config['bags'][i]['segments'][j]['topics'] = [self.config['bags'][i]['segments'][j]['topics']]
```
Comment (DSsoto) [inline utils/mil_tools/mil_ros_tools/label_me_bag.py]: Why not:
```python
segment = [segment]
```
Does that not work?

**Review by DSsoto — APPROVED** [review]
Good work man, this is pretty useful.

**Comment by DSsoto** [comment]
Will wait for @whispercoros's review before merging.

File: perception/mil_vision/mil_vision_tools/rect_finder.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        rvec to a 3x3 rotation matrix.
+        '''
+        corners = np.array(corners,dtype=np.float)
+        # Use camera intrinsics and knowledge of marker's real demensions to get a pose estimate in camera frame
+        if cam is not None:
+            intrinsics = cam.intrinsicMatrix()
+            if rectified:
+                dist_coeffs = np.zeros((5,1))
+            else:
+                dist_coeffs = distortionCoeffs()
```
Comment (kev-the-dev) [inline perception/mil_vision/mil_vision_tools/rect_finder.py]: Oops should be ```cam.distortionCoeffs()```

File: utils/mil_tools/mil_ros_tools/labeled_bag_tester.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        for topic, msg, t in self._cur_bag.read_messages(topics=self._pubs.keys()):
+            if msg._type == 'mil_msgs/LabeledObjects':
+                self.verify(msg)
+            x = time.time()
+            if rospy.is_shutdown():
+                return
+            dt = t - last_time
+            last_time = t
+            self._pubs[topic].publish(msg)
+            dx = time.time() - x
```
Comment (DSsoto) [inline utils/mil_tools/mil_ros_tools/labeled_bag_tester.py]: Why not use rostime? This prevents you from running this at different bag playback speeds.

**Review by DSsoto — CHANGES_REQUESTED** [review]
This looks good. How was this tested? Requesting some usage documentation for the labeled bag tester before merging.

**Comment by kev-the-dev** [comment]
I'm getting rid of labeled_bag_tester in this PR as it's still WIP and I'd rather get the other stuff merged in the meantime. The rect_finder was tested by replacing the relevant parts of path_marker_finder in subjugator (which will be PRed once this is merged).

File: perception/mil_vision/mil_vision_tools/rect_finder.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+    def to_polygon(self):
+        '''
+        Returns a geometry_msgs/Polygon representing the four corners
+        of the rectangle where all z = 0
+        '''
+        return numpy_to_polygon(self.model_3D)
+
+    @staticmethod
+    def sort_corners(rect, debug_image=None):
```
Comment (DSsoto) [inline perception/mil_vision/mil_vision_tools/rect_finder.py:67]: Thanks for doing this! Good work!

File: perception/mil_vision/mil_vision_tools/rect_finder.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    def verify_contour(self, contour):
+        '''
+        Returns a numerical comparison between a contour and the perfect
+        model of the rectangle. A perfectly matched contour returns 0.0.
+
+        Useful for filtering contours.
+        '''
+        return cv2.matchShapes(contour, self.model_2D, 3, 0.0)
+
+    def get_corners(self, contour, epsilon_factor=0.02, debug_image=None):
```
Comment (DSsoto) [inline perception/mil_vision/mil_vision_tools/rect_finder.py:114]: One way I reliably approximated contours as quadrilaterals was by doing it iteratively and increasing epsilon until I had a quadrilateral. Its more robust for when over-saturation or caustics make it hard to retrieve the quadrilateral. 

Check [this](https://github.com/uf-mil/SubjuGator/blob/5d169e9d23d6b20ef488040257fc0837ac887adb/perception/sub8_perception/nodes/torpedo_board.cpp#L1095) out for an implementation.

File: perception/mil_vision/mil_vision_tools/rect_finder.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        '''
+        epsilon = epsilon_factor*cv2.arcLength(contour,True)
+        polygon = cv2.approxPolyDP(contour,epsilon,True)
+        if len(polygon) != 4:
+            return None
+        return self.sort_corners(polygon, debug_image=debug_image)
+
+    def get_pose_3D(self, corners, intrinsics=None, dist_coeffs=None, cam=None, rectified=False):
+        '''
+        Uses the model of the object, the corosponding pixels in the image, and camera
```
Comment (DSsoto) [inline perception/mil_vision/mil_vision_tools/rect_finder.py]: `corresponding`

File: perception/mil_vision/mil_vision_tools/rect_finder.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+        @return a tuple (center, vecter)
+        '''
+        top_center = (corners[1]+corners[0])/2.0
+        bot_center = (corners[2]+corners[3])/2.0
+        vector = top_center - bot_center
+        vector = vector / np.linalg.norm(vector)
+        center = bot_center + (top_center - bot_center)/2.0
+        vector = top_center - bot_center
+        vector = vector/np.linalg.norm(vector)
```
Comment (DSsoto) [inline perception/mil_vision/mil_vision_tools/rect_finder.py]: This will fail flake8 checks soon.

**Review by DSsoto — APPROVED** [review]
Dude great work with RectFinder! Thats some good code. Oooohwieeee!

The bag stuff is good too. Next time, I would make these kind of things separate PR's though.

Let me know if you want me to merge or you want to take my more robust quadrilateral approximation suggestion.

File: perception/mil_vision/mil_vision_tools/rect_finder.py (hunk 5 above)
Comment (kev-the-dev) [reply perception/mil_vision/mil_vision_tools/rect_finder.py:114]: Yeah I was having problems with that, great idea! I'll implement that in a future PR.

File: perception/mil_vision/mil_vision_tools/rect_finder.py (hunk 7 above)
Comment (kev-the-dev) [reply perception/mil_vision/mil_vision_tools/rect_finder.py]: I corrected all the issues outputted by flake8 in all the files changed in this PR

**Pull request merged**